R/mod_load_data.R

```diff
@@ -122,28 +160,143 @@ mod_load_data_server <- function(input, output, session, r) {
     req(r$data, input$user_DT_columns_selected)
 
+    #TODO - this assumes Year and Quarter are in cols 1:2.
     cols_to_keep <- input$user_DT_columns_selected
     if (any(c(1, 2) %not_in% cols_to_keep)) {
-      warning("Dropping Year or Quarter is a bad idea so lets not.")
+      warning("Dropping Year or Quarter is a bad idea so let's not.")
       cols_to_keep <- union(c(1, 2), cols_to_keep)
     }
 
+    # store a copy of the data pre-change so that user can undo one step
     r$data_old <- r$data
-    r$data <- r$data %>% select(1:2, cols_to_keep)
+
+    # keep selected columns only
+    r$data <- r$data %>% dplyr::select(Year, Quarter, cols_to_keep)
+
+    # update undifferenced copy for reference
+    r$data_undiff <- r$data_undiff[, cols_to_keep]
+  })
+
+  # when user updates last period of historical data, setup end period for
+  # models
+  observeEvent(input$RangeHistorical, {
+    req(r$data, r$dep_var)
+
+    # read in the bounds for historical data (from user input)
+    r$date_start <- lubridate::yq(input$RangeHistorical[1])
+    r$date_start <- c(lubridate::year(r$date_start), lubridate::quarter(r$date_start))
+    r$date_end <- lubridate::yq(input$RangeHistorical[2])
+    r$date_end <- c(lubridate::year(r$date_end), lubridate::quarter(r$date_end))
   })
 
+  # slider input - Historical data -----
+  output$RangeHistorical <- renderUI({
+    req(r$data, r$dep_var)
+    df <- r$data
+
+    # create the sequence of Date objects
+    dateList <- seq(lubridate::yq(paste0(df[1, "Year"],
+                                         ": Q",
+                                         df[1, "Quarter"])),
+                    to = lubridate::yq(paste0(df[nrow(df), "Year"],
+                                              ": Q",
+                                              df[nrow(df), "Quarter"])),
+                    by = "quarter")
+
+    # format vector
+    dateListFormatted <- zoo::as.yearqtr(dateList)
+
+    # find default end for historical data
+    # (based on when dependent variable ends)
+    if (is.null(r$dep_var)) {
+      defaultEnd <- dateListFormatted[length(dateListFormatted)]
+    } else {
+      # find the last data point for the selected dependent variable
+      lastDepVarDataPoint <- df %>%
+        select("Year", "Quarter", r$dep_var)
+      lastDepVarDataPoint <- zoo::na.trim(lastDepVarDataPoint)
+      lastDepVarDataPoint <- lastDepVarDataPoint[nrow(lastDepVarDataPoint), ]
+
+      defaultEnd <- c(as.numeric(lastDepVarDataPoint[, "Year"]),
+                      as.numeric(lastDepVarDataPoint[, "Quarter"]))
+    }
+
+    # put together widget
+    shinyWidgets::sliderTextInput(
+      inputId = ns("RangeHistorical"),
+      label = "Select the start and end points for the historical data",
+      grid = TRUE,
+      force_edges = TRUE,
+      choices = dateListFormatted,
+      selected = c(dateListFormatted[1],
+                   defaultEnd)
+    )
+  })
+
+  # # slider input - Projection data -----
```

Comment:
I'm not sure we should be committing commented our code to the final main branch.